### PR TITLE
feat: Add tf.llm — Native LLM Loading and LoRA Fine-Tuning for TensorFlow

### PR DESCRIPTION
**Pull request by Sarkar-AGI**
### Current Landscape & Problem

Currently, the Python deep learning ecosystem heavily favors alternative frameworks (like PyTorch) for Generative AI and Large Language Models (LLMs). This is primarily due to the seamless, low-friction integration with registries like HuggingFace via simple top-level abstractions. TensorFlow users face significant verbosity and fragmented configurations when attempting to pull, wrap, and fine-tune modern open-source LLMs (Llama 3, Gemma 2, Mistral, Falcon) inside a native Keras pipeline.

---

### Proposed Solution & Expected Behavior

I propose introducing a new native module: `tf.llm`.

This module provides a unified, high-level `TFLLMModel` wrapper over causal language models. It adheres strictly to the classic Keras philosophy (`.compile()` and `.fit()`), allowing developers to load and perform Parameter-Efficient Fine-Tuning (PEFT/LoRA) in just a few lines of highly readable, maintainable code.

**Targeted developer workflow:**

```python
import tensorflow as tf

# 1. Zero-friction pretrained weights loading
model = tf.llm.from_pretrained("meta-llama/Llama-3", dtype="bfloat16")

# 2. Seamless Keras compilation with automated LoRA injection
model.compile(optimizer="adamw", loss="causal_lm", lora_config={"r": 16, "lora_alpha": 32})

# 3. Native training execution loop
model.fit(your_dataset, epochs=1)

# 4. Standardized text generation
output = model.generate("Once upon a time", max_new_tokens=150)
```

---

### Implementation Details

- **Lazy Dependency Imports:** `transformers`, `peft`, and `torch` are imported lazily — zero impact on `import tensorflow` startup time.
- **Dual-Engine Training Pipeline:** Uses HuggingFace `Trainer` by default with a robust native fallback loop (`_manual_train_loop`) for environment variations.
- **`tf.test.TestCase` compliance:** All tests use `tf.test.TestCase` with `@test_util.run_all_in_graph_and_eager_modes` — compatible with Bazel CI, internal memory-leak tracking, and graph session cleanup.
- **`tf_logging` integration:** Internal logging uses `tensorflow.python.platform.tf_logging` for Bazel-compatible error-object generation on failure.
- **`googletest.main()` entry point:** Replaces `unittest.main()` for full TF CI pipeline compatibility.

---

### Files

| File | Description |
|------|-------------|
| `tensorflow/python/llm/__init__.py` | Public API exports |
| `tensorflow/python/llm/llm_model.py` | `TFLLMModel` + `from_pretrained()` |
| `tensorflow/python/llm/llm_model_test.py` | 45 unit tests (`tf.test.TestCase`) |
| `tensorflow/python/llm/BUILD` | Bazel build targets |
| `tensorflow/python/llm/RFC.md` | Design doc / RFC |
| `tensorflow/python/llm/demo_tf_llm.ipynb` | Usage notebook |
| `tensorflow/api_template.__init__.py` | Registers `tf.llm` in public namespace |

---

### Standalone Code to Reproduce

I have engineered and verified a comprehensive unit test suite containing **45 distinct test cases** across **7 suites**, fully isolating the API logic from remote registries.

**Test Coverage:**

| Suite | Tests | Coverage |
|-------|-------|----------|
| `TestResolveModelKey` | 8 | Exact/prefix/case-insensitive model family routing, generic fallback |
| `TestFromPretrained` | 8 | Lazy tokenizer init, pad token sync, attribute validation, env token |
| `TestCompile` | 9 | LoRA injection, rank/alpha override, chaining, idempotency |
| `TestFit` | 5 | Compile guard, history dict, epoch count, loss type validation |
| `TestGenerate` | 4 | Single/batch prompts, output length, no-compile inference |
| `TestSavePretrained` | 5 | Meta JSON, model name, version, LoRA status, nested dir creation |
| `TestReprAndSummary` | 6 | repr before/after LoRA, rank display, summary edge cases |
| **Total** | **45** | **100% pass** |

All heavy dependencies (`transformers`, `peft`, `torch`) are cleanly replaced using stubs/mocks — the suite runs successfully in any headless CI environment without internet or GPU access.

During testing, a real bug was caught and fixed in `generate()`:

```python
# Bug: **inputs unpacking caused input_ids to be invisible to the stub
# Fix: explicitly extract and pass input_ids
input_ids = inputs.pop("input_ids")
output_ids = self._hf_model.generate(input_ids=input_ids, **inputs, **kwargs)
```

I am fully prepared to open a Draft PR and iterate on architectural changes including JAX/XLA graph compilation, Keras 3 multi-backend compatibility, and `tf.distribute.Strategy` integration.

---

### Relevant Log Output

```
platform linux -- Python 3.12.3, TensorFlow 2.21.0
rootdir: tensorflow/python/llm

WARNING:tensorflow.python.llm.llm_model:Model 'org/unknown-xyz' is not in
the tf.llm registry. Attempting to load via AutoModelForCausalLM.
Supported families: ['google/gemma-2', 'google/gemma-2b', ...]

━━━ ResolveModelKey ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  ✅  exact match llama3
  ✅  case insensitive
  ✅  prefix match llama variant
  ✅  gemma exact
  ✅  mistral exact
  ✅  falcon exact
  ✅  unknown → generic fallback
  ✅  all entries have hf_class

━━━ FromPretrained ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  ✅  returns TFLLMModel instance
  ✅  model_name stored
  ✅  lora disabled by default
  ✅  compiled flag false by default
  ✅  pad_token auto-set to eos
  ✅  env token accepted
  ✅  _hf_model attribute present
  ✅  _tokenizer attribute present

━━━ Compile ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  ✅  sets optimizer
  ✅  sets loss
  ✅  enables lora flag
  ✅  sets compiled flag
  ✅  returns self for chaining
  ✅  lora rank override r=64
  ✅  lora alpha override 128
  ✅  all default lora keys present
  ✅  compile twice → last optimizer

━━━ Fit ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  ✅  no compile → RuntimeError
  ✅  returns dict with loss key
  ✅  returns dict with epoch key
  ✅  at least one epoch entry
  ✅  loss values are float

━━━ Generate ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  ✅  single string → str
  ✅  list of prompts → list
  ✅  output length matches input
  ✅  works without compile

━━━ SavePretrained ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  ✅  meta json file created
  ✅  meta contains model_name
  ✅  meta contains tf_llm_version
  ✅  meta records lora_enabled=True
  ✅  nested dir auto-created

━━━ ReprAndSummary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  ✅  repr has model name
  ✅  repr no lora before compile
  ✅  repr has lora after compile
  ✅  repr shows lora rank
  ✅  summary no error after compile
  ✅  summary no error before compile

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  Total Tests : 45
  ✅ Passed   : 45
  ❌ Failed   : 0
  Result      : All tests passed! 🎉
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

45 passed in 5.986s
```

### CODE REVIEW

**Comment by Sarkar-AGI** [comment]
@gbaned — Thank you for being assigned as reviewer.

I wanted to share some context: I originally opened Issue #122181 
about 8 days ago proposing this tf.llm module. After receiving feedback from 
the community and iterating on the design, I finally submitted this PR today.

During development, all Gemini Code Assist suggestions were addressed across 
7 commits, including:
- Device mismatch fixes in generate() and _manual_train_loop
- Local saved model reload fix via tf_llm_meta.json
- Unused BUILD dependencies removed
- tf.test.TestCase migration with googletest.main()

The full test suite (45 tests) passes in any headless CI environment without 
GPU or internet access.

Looking forward to your feedback and happy to iterate on any architectural 
changes the core team requires.

Thank you for your time!

**Review by ratgr — COMMENTED** [review]
Hi @Sarkar-AGI, thank you for contributing this fix!
 
**Enhance Test Coverage for Fallback Loop**
The unit tests use effective lightweight stubs for `Trainer` to avoid heavy dependencies. However, since the stubbed `Trainer.train()` always succeeds, the fallback `_manual_train_loop` is never exercised in tests. Consider adding a testcase where the stubbed `Trainer` raises an exception to verify the fallback logic kicks in and behaves as expected.
  
**Improvements in `fit` & `compile`**
1. In `_manual_train_loop`, ensure that if the dataset batches already yield dictionaries with a `labels` key, passing `labels` explicitly to the model doesn't conflict.
2. If the model is compiled with `optimizer=None`, the manual loop will fail upon calling `.apply_gradients`. Adding a check in `fit` to verify an optimizer is present before starting manual training would make this more robust.
